### PR TITLE
Enable shortcut while CommandPalette is open

### LINKED ISSUE #6679: `commandPalette` keybinding should close the palette when it's open
This is being moved from discussion in #6635.

We wanted to enable the `commandPalette` keybinding to also work within the palette itself. 

carlos-zamora
> I know I suggested this, but it unveiled a huge problem. When paste is bound to ctrl+v, we paste to the underlying terminal twice. A similarly weird problem occurs with copy.
> 
> Maybe we should only detect the commandPalette action and use that as an additional way to dismiss it.

DHowett
> Don't we just need to mark it Handled here? lol

zadjii-msft
> This is shockingly bad actually.
> 
> * If we just mark it as Handled(true), then the following KeyUp will still land in the TermControl, resulting in ^VPasted Text being written to the Terminal.
> * If we change the keybinding handler to on KeyUp, then it's likely that the first keyup it'll see is the keybinding for toggle the palette, which will immediately close it.
> 
> I'm going to move this to a follow-up issue, to get at the least the toggle the command palette action back, and maybe the full keybindings

> [3:29 PM] Mike Griese
>     I'll file a follow-up - maybe we could do something where the action  is selected on the keydown, but then only dispatched on the keyup?
​> [3:30 PM] Mike Griese
>    so we'll only try a keybinding on a keyup after a non-modifier keydown

### LINKED ISSUE #7914: [Functional-Command Palette]- User is not able to perform action using shortcuts when command palette is open.
**User Experience:**
Users who rely on screen reader/Keyboard will not be able to perform action efficiently if shortcuts are not working when command palette is open

**Test Environment:**
App: Windows Terminal Preview
Feature: Command Palette
OS: Win 10 2004 Build (19041.508)

**Repro Steps:**
1.Open Windows terminal Preview app
2.Open Command Palette using shortcut (Ctrl+shift+p)
3.Verify whether user is able to perform any shortcut operation when command palette is open. 

**Actual:**
User is not able to perform action using shortcuts when command palette is open.

**Expected:**
User should be able to perform the shortcut operation even when the command palette is open.

### PR DESCRIPTION
**Pull request by Hegunumo**
## Summary of the Pull Request
It is maybe not the best way since I had to get all the cases for key handling so I just created for each of them. As a result the code get longer(not optimized). Most difficult thing was Next tab and Previous tab I just could not solve it.

### 9 commands that couldn't enabled > <
     1. Increase font size               -> could not find VirtualKey for "-"
     2. Decrease font size             -> could not find VirtualKey for "+"
     3. Split pane, split:horizontal -> could not find VirtualKey for "-"
     4. Split pane, split:vertical     -> could not find VirtualKey for "+"
     5. Open default settings        -> could not find VirtualKey for ","
     6. Open settings file               -> could not find VirtualKey for ","
     7. Open new tab dropdown    -> could not resolve keybindings
     8. Next tab                               -> could not resolve keybindings
     9. Previous tab                        -> could not resolve keybindings
 
 
## PR Checklist
* [x] Closes #6679
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [ ] Tests added/passed
* [ ] Documentation updated. If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/terminal) and link it here: #xxx
* [ ] Schema updated.
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

### CODE REVIEW

**Comment by Hegunumo** [comment]
I will try 🙃

**Comment by Hegunumo** [comment]
I'm trying to get default.json using `CascadiaSettings::LoadDefaults()` and do `KeyMap()`, but it dosen't work.

**Comment by zadjii-msft** [comment]
> 
> 
> I'm trying to get default.json using `CascadiaSettings::LoadDefaults()` and do `KeyMap()`, but it dosen't work.


I wouldn't bother with all that - the `TerminalPage` already has a perfectly good `CascadiaSettings` object (`_settings`) that's got the entirely loaded settings. Don't bother re-loading the settings, just use that one.

**Comment by Hegunumo** [comment]
Should I get `IKeyBindings` as input for _keyDownHandler() in CommandPalette.cpp ?
It says function dose not take 2 arguments.

File: src/cascadia/TerminalApp/CommandPalette.cpp [hunk 1]
```diff
@@ -162,6 +165,22 @@ namespace winrt::TerminalApp::implementation
...
+        for (int vKey = 0; vKey != 219; vKey++)
+        {
+            VirtualKey virtualKey = static_cast<VirtualKey>(vKey);
+            if (virtualKey == key)
+            {
+                kc = { ctrl, alt, shift, static_cast<int32_t>(virtualKey) };
+                return kc;
+            }
+        }
+        return kc = nullptr;
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/CommandPalette.cpp:181]: Why isn't this whole function just 
```suggestion
        return winrt::Microsoft::Terminal::TerminalControl::KeyChord{ctrl, alt, shift, static_cast<int32_t>(key)};
```

File: src/cascadia/TerminalApp/CommandPalette.cpp [hunk 2]
```diff
@@ -225,20 +244,51 @@ namespace winrt::TerminalApp::implementation
...
+        }
+        else if (shiftDown && altDown)
+        {
+            kc = SetKeyChord(false, true, true, key);
+        }
+        else if (shiftDown && altDown && ctrlDown)
+        {
+            kc = SetKeyChord(true, true, true, key);
+        }
+
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/CommandPalette.cpp:280]: Why bother with all this, branching? Wouldn't 
```c++
kc = SetKeyChord(ctrlDown , altDown, shiftDown , key);
```
be easier?

File: src/cascadia/TerminalApp/CommandPalette.cpp [hunk 3]
```diff
@@ -225,20 +244,51 @@ namespace winrt::TerminalApp::implementation
...
+        auto setting = ::winrt::TerminalApp::implementation::AppLogic::CurrentAppSettings();
+        auto keymap = setting.GlobalSettings().KeyMap();
+        const auto action = keymap.TryLookup(kc);
+        if (action)
+        {
+            _close();
+            _dispatch.DoAction(action);
             e.Handled(true);
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/CommandPalette.cpp:288]: Should we early `return` now that the even was handled?

**Review by zadjii-msft — CHANGES_REQUESTED** [review]

**Review by zadjii-msft — CHANGES_REQUESTED** [review]
This works pretty well actually! There's only two cases that seem weird, and I'm not really sure if they're worth blocking this whole PR over:
* If you press <kbd>ctrl+shift+p</kbd> to dismiss the command palette, then immediately the palette will re-open itself on the key up. Maybe we need to special-case that action, to only dispatch it on a key-up?
* <kbd>ctrl+tab</kbd> to move to the next tab doesn't actually dispatch that action. It seems like the text box eats that keystroke first. That's probably okay though? At least intuitively, I saw that happen and thought "oh, yea that makes sense".

So I'm blocking over the `ToggleCommandPalette` action thing, but I am open to being convinced otherwise by other reviewers

**Comment by Hegunumo** [comment]
@zadjii-msft I fixed second issue, but I couldn't understand first problem. When I press `ctrl+shift+p` the cursor blinks, I couldn't even dismiss the command palette. Dose this also occur for you as well?

**Comment by zadjii-msft** [comment]
Yea, so that's basically what's happening to me to. I'm guessing that what's happening is that all in the course of 1-2 frames, the palette is being dismissed, the focus is moving back to the control (causing the cursor to blink), and then on the key up of <kbd>P</kbd>, we're sending _another_ `toggleCommandPalette` action, which is causing the palette to re-appear. 

I don't have the branch checked out locally right now to confirm, but I bet if you press and hold the <kbd>P</kbd> key, you won't see the palette re-appear until the key is released.

**Comment by Hegunumo** [comment]
I forgot that I was closing CommandPalette at first. Now I changed to close it when action is not `toggleCommandPalette`.

File: src/cascadia/TerminalApp/CommandPalette.cpp [hunk 4]
```diff
@@ -211,6 +212,14 @@ namespace winrt::TerminalApp::implementation
...
             }
         }
+        if (_currentMode == CommandPaletteMode::ActionMode && key == VirtualKey::Tab)
+        {
+            auto const state = CoreWindow::GetForCurrentThread().GetKeyState(winrt::Windows::System::VirtualKey::Control);
+            if (WI_IsFlagSet(state, CoreVirtualKeyStates::Down))
+            {
+                _keyDownHandler(sender, e);
+            }
+        }
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/CommandPalette.cpp:222]: I think we should probably remove this special case - I think there's a bunch of other similar keystrokes that won't work _exactly_ right, but at least this makes intuitive sense when it doesn't work. 

Sorry if I was confusing before!

**Review by zadjii-msft — CHANGES_REQUESTED** [review]
Sorry for misleading you on the <kbd>ctrl+tab</kbd> thing - let's remove that special case handling, but otherwise this feels good. Thanks!

File: src/cascadia/TerminalApp/CommandPalette.cpp (hunk 4 above)
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/CommandPalette.cpp:222]:
```suggestion
```

File: src/cascadia/TerminalApp/CommandPalette.cpp [hunk 5]
```diff
@@ -185,7 +186,7 @@ namespace winrt::TerminalApp::implementation
         }
     }
 
-    void CommandPalette::_previewKeyDownHandler(IInspectable const& /*sender*/,
+    void CommandPalette::_previewKeyDownHandler(IInspectable const& sender,
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/CommandPalette.cpp:189]:
```suggestion
    void CommandPalette::_previewKeyDownHandler(IInspectable const& /*sender*/,
```

**Review by zadjii-msft — APPROVED** [review]
Thanks!

File: src/cascadia/TerminalApp/CommandPalette.cpp [hunk 6]
```diff
@@ -284,7 +285,23 @@ namespace winrt::TerminalApp::implementation
     {
         auto key = e.OriginalKey();
         auto const ctrlDown = WI_IsFlagSet(CoreWindow::GetForCurrentThread().GetKeyState(winrt::Windows::System::VirtualKey::Control), CoreVirtualKeyStates::Down);
+        auto const altDown = WI_IsFlagSet(CoreWindow::GetForCurrentThread().GetKeyState(winrt::Windows::System::VirtualKey::Menu), CoreVirtualKeyStates::Down);
+        auto const shiftDown = WI_IsFlagSet(CoreWindow::GetForCurrentThread().GetKeyState(winrt::Windows::System::VirtualKey::Shift), CoreVirtualKeyStates::Down);
```
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/CommandPalette.cpp:289]: You could probably remove lines 370-372 now?

**Comment by Jspooner09** [comment]
I have a virus in computer


Jimmy R. 

CONFIDENTIALITY: The information contained in this transmission may contain privileged and confidential information. It is intended only for the use of the person(s) named above. If you aren not the intended recipient, you are hereby notified that any review, dissemination, dipsltribution or duplication of this communication, and the information contained in it, is strictly prohibited. If you are not the intended recipient, please contact the sender and immediately destroy all copies of the original message.



On Thursday, December 3, 2020, 3:38 PM, Carlos Zamora <notifications@github.com> wrote:




@carlos-zamora commented on this pull request.

In src/cascadia/TerminalApp/CommandPalette.cpp:
> @@ -284,7 +285,23 @@ namespace winrt::TerminalApp::implementation
     {
         auto key = e.OriginalKey();
         auto const ctrlDown = WI_IsFlagSet(CoreWindow::GetForCurrentThread().GetKeyState(winrt::Windows::System::VirtualKey::Control), CoreVirtualKeyStates::Down);
+        auto const altDown = WI_IsFlagSet(CoreWindow::GetForCurrentThread().GetKeyState(winrt::Windows::System::VirtualKey::Menu), CoreVirtualKeyStates::Down);
+        auto const shiftDown = WI_IsFlagSet(CoreWindow::GetForCurrentThread().GetKeyState(winrt::Windows::System::VirtualKey::Shift), CoreVirtualKeyStates::Down);
+
+        winrt::Microsoft::Terminal::TerminalControl::KeyChord kc{ ctrlDown, altDown, shiftDown, static_cast<int32_t>(key) };
+        auto setting = ::winrt::TerminalApp::implementation::AppLogic::CurrentAppSettings();
⬇️ Suggested change-        auto setting = ::winrt::TerminalApp::implementation::AppLogic::CurrentAppSettings();
+        auto setting { AppLogic::CurrentAppSettings()};

nit. We're already in the winrt::TerminalApp::implementation namespace

In src/cascadia/TerminalApp/CommandPalette.cpp:
> @@ -284,7 +285,23 @@ namespace winrt::TerminalApp::implementation
     {
         auto key = e.OriginalKey();

mind updating the method description at the top of this function too

In src/cascadia/TerminalApp/CommandPalette.cpp:
>  
+        if (action)
+        {
+            if (action.Action() != ShortcutAction::ToggleCommandPalette)
+            {
+                _close();
+            }
+            _dispatch.DoAction(action);
+            return;
+        }
         if (key == VirtualKey::Up)
⬇️ Suggested change-        if (key == VirtualKey::Up)
+        else if (key == VirtualKey::Up)

?

In src/cascadia/TerminalApp/CommandPalette.cpp:
>  
+        if (action)
+        {
+            if (action.Action() != ShortcutAction::ToggleCommandPalette)
+            {
+                _close();
+            }
+            _dispatch.DoAction(action);
+            return;

Should we mark e as Handled too?

In src/cascadia/TerminalApp/CommandPalette.cpp:
>          auto const ctrlDown = WI_IsFlagSet(CoreWindow::GetForCurrentThread().GetKeyState(winrt::Windows::System::VirtualKey::Control), CoreVirtualKeyStates::Down);
+        auto const altDown = WI_IsFlagSet(CoreWindow::GetForCurrentThread().GetKeyState(winrt::Windows::System::VirtualKey::Menu), CoreVirtualKeyStates::Down);
+        auto const shiftDown = WI_IsFlagSet(CoreWindow::GetForCurrentThread().GetKeyState(winrt::Windows::System::VirtualKey::Shift), CoreVirtualKeyStates::Down);

You could probably remove lines 370-372 now?

In src/cascadia/TerminalApp/CommandPalette.cpp:
>  
+        if (action)
+        {
+            if (action.Action() != ShortcutAction::ToggleCommandPalette)
+            {
+                _close();
+            }
+            _dispatch.DoAction(action);
+            return;
+        }
         if (key == VirtualKey::Up)

This would also let us get rid of the return; above, if this works right.

**Comment by Hegunumo** [comment]
@carlos-zamora I made a change. Please review!

**Review by carlos-zamora — APPROVED** [review]
Thanks!

**Comment by DHowett** [comment]
@carlos-zamora when you automerge, please try to make sure you edit the PR body to explain what actually happened. It looks like the words in the body of this PR are way, way different from the final implementation! Future readers who only have the git blame will not have the context y’all had when reviewing it.

**Comment by DHowett** [comment]
Hi @Hegunumo! Sorry about this. I had to revert this pull request because it caused a build break in `main`.

Reversion: 61ccf44f224356858664d3fa00d8342416d2a888
Reopening #6679

If you make a fix on top of commit 1965fb5, we can merge the two back together.

**Pull request merged**